data_transfer/devices/dreem.py

```diff
@@ -46,41 +62,116 @@ def download_metadata(self) -> None:
+
+                # Best-case: only one device was worn and UCAM knows it
+                if len(dreem_devices) == 1:
+                    device_record = dreem_devices[0]
+                # Edge-case: multiple dreem headbands used, e.g., if one broke.
+                elif len(dreem_devices) > 1:
+                    # Determine usage based on weartime as it exists in UCAM
+                    device_record = ucam.record_by_wear_period_in_list(
+                        dreem_devices, recording.start, recording.end
+                    )
+                # Edge-case: device not logged with patient in UCAM
+                else:
+                    print(f"Metadata cannot be determined for:\n    {recording}")
+                    continue
+            else:
+                unknown += 1
+                print(f"Metadata cannot be determined for:\n    {recording}")
+                continue
 
-            start_wear = utils.format_weartime(history["checkout"], "inventory")
-            # NOTE: if device not returned the current day is used.
-            end_wear = utils.format_weartime(history["checkin"], "inventory")
             record = Record(
-                # NOTE: id is a unique uuid used to GET raw data from Dreem
-                filename=item["id"],
-                device_type=utils.DeviceType.DRM.name,
-                device_id=device_id,
+                filename=recording.id,
+                device_type=dtype,
                 patient_id=patient_id,
-                start_wear=start_wear,
-                end_wear=end_wear,
+                **asdict(device_record),
             )
 
             create_record(record)
+            print(f"Record Created:\n   {record}")
 
             path = Path(config.storage_vol / f"{record.filename}-meta.json")
             # Store metadata from memory to file
             utils.write_json(path, item)
 
+            print(f"Metadata saved to: {path}\n")
+        print(f"{known} records created and {unknown} NOT this session.")
+
+    def __recording_metadata(self, recording: dict) -> DreemRecording:
+        """
+        Maps data from Dreem response to class to simplify access/logging.
+        """
+        id = recording["id"]
+        device_id = recording["device"]
+        user_id = recording["user"]
+        # When the recording took place
+        start = datetime.fromtimestamp(recording["report"]["start_time"])
+        end = datetime.fromtimestamp(recording["report"]["stop_time"])
+
+        return DreemRecording(*(id, device_id, user_id, start, end))
+
+    def __patient_id_from_ucam(
+        self, device_serial: str, dreem_start: datetime, dreem_end: datetime
+    ) -> Optional[str]:
+        """
+        Determine PatientID by wear period of device in UCAM.
+        """
+        # NOTE/TODO: given this is a 1-1 mapping, why not use a local CSV?
+        device_id = inventory.device_id_by_serial(device_serial)
```

Comment:
Might be worthwhile to convert the `inventory.device_id_by_serial` API to read from a CSV since each device will be assigned one Serial ID and this is unlikely to change? This would save us making many external API calls to the inventory. Alternatively, we should look into caching the results from the inventory to prevent similar requests as this.

Reply:
Agreed. With a number of `.csv` file we work from, might be good to add a folder in the repo.

Reply:
Actually, how many Dreem devices do you foresee we having in our inventory? Otherwise, we could alternatively add another query in the  _consumer > routers > inventory_, such as `/device/bytype/{type}` with type being the Dreem device. 

At the start of the Dreem DAG, this would the query `snipeit` once, e.g.
```shell 
GET snipeit_url/hardware?limit=500&model_id=8'
```
where model = 8 corresponds to the Dreem Device I believe. If you think more than 500 Dreem devices might be present, we could increase the number of items by increasing [MAX_RESULTS in our .env file](150 requests by default). We could even further filter the results by adding a `location_id` parameter Thoughts?

Reply:
> Actually, how many Dreem devices do you foresee we having in our inventory? Otherwise, we could alternatively add another query in the consumer > routers > inventory, such as /device/bytype/{type} with type being the Dreem device.

Currently **46** and potentially at most **200** in the future.  Great find @davidverweij -- not sure how I missed it! We might be able to filter by `serial` as a GET parameter too -- I will have a look. 

> If you think more than 500 Dreem devices might be present, we could increase the number of items by increasing [MAX_RESULTS in our .env file](150 requests by default). We could even further filter the results by adding a location_id parameter Thoughts?

The bigger issue is that even though we have 40 devices, we could make potentially 3 API calls to inventory per data record, so we would hit the threshold of 150 requests quite easily. A caching solution for the same queries would be another potential solution since the queries are often the same. For example, for dreem if there are 10 data recordings created on the same device, we might have to lookup the DeviceID by serial 10 times.

Reply:
I've referenced this discussion on the [issue on the stack](https://github.com/ideafast/stack/issues/9#issuecomment-781258338)